IPIP/0298-denylists-for-ipfs-nodes-and-gateways.md

```diff
@@ -0,0 +1,243 @@
+```
+
+### List Provider
+
+A List Provider is an entity building and providing a list of content to be allowed/denied. They can provide their lists through the IPFS network or via [an HTTP Gateway endpoint](../http-gateways/README.md).
+
+### IPFS Node
+
+An IPFS node is a service that interacts with the IPFS network. This could be because it's providing content to the network, because it's retrieving content, or because it's serving IPFS hosted content through an HTTP interface.
+
+To interact with content lists, it operates a rule engine. The rule engine decides what action should be taken for a particular request.
+For instance, if CID bafy...badcid is denied, the bitswap engine should not request it, and access to bafy...badcid through a node exposed gateway are going to 403 forbidden.
+
+### User
+
+User is an actor that can make request to the IPFS Node. It's considered authenticated already.
+Access to the node can be made via an HTTP gateway, Bitswap, or another transport method exposed by the IPFS Node.
+
+It is understood that the User might be in control of the IPFS Node, and is able to configure the rule engine as it suits them.
+
+## Schemas
+
+### Definition
+
+The schemas below are defined as [IPLD schemas](https://ipld.io/specs/schemas/).
+
+```ipldsch
+type List struct {
+  Action Int # Suggested action to be taken
+  StatusCode optional Int # HTTP status code to be returned by IPFS gateways by default
+
+  Entries optional Entry[] # Entries included in the list
+  Imports optional Bytes # CID to a chain of entries included in the list
+
+  Description optional String # Context to be shared with the list consumers
+}
+
+type Entry struct {
+  Type Int # Type of the content string. Refer to CID_V1, ABSOLUTE_PATH, PARTIAL_PATH, SHA256_CID_V1
+  Content String # Encoded content based on the Type
+  Description optional String # Context to be shared with the entry consumers
+  StatusCode optional Int  # HTTP status code to be returned by IPFS gateways for this specific Entry
+}
+
+type EntryImport struct {
+  Entries Bytes[] # List of CID associated to Entry objects
+  Next nullable Bytes # CID for the next EntryImport object
+}
+```
+
+### Example
+
+Here is a denylist in the proposed format.
+
+```javascript
+{
+  Action: 2, // Action.BLOCK
+  StatusCode: 451,
+  Description: "Denylist of ACME Corp. on 1970/01/01",
+  Entries: [
+    {
+      Type: 0, // Type.CID_V1
+      Content: "bafybeihfqymzmqhbutdd7i4mkq2ltzznzgoshi4r2pnv4hsc2acsojawoe",
+      Description: "ipfs quick start"
+    },
+    {
+      Type: 1, // Type.ABSOLUTE_PATH
+      Content: "/ipns/bad.example.com",
+      Description: "Bad path"
+    },
+    {
+      Type: 2, // Type.PARTIAL_PATH
+      Content: "bag.again.example.com",
+      StatusCode: 410
+    },
+    {
+      Type: 3, // Type.SHA256_CID_V1
+      Content: "9056e0f9948c942c16af3564af56d4bb96b6203ad9ccd3425ec628bcd843cc39",
+      Description: "sensitive cid that needs to be blocked"
+    }
+  ]
```

Comment:
Note that it is very difficult to append anything to this list. You cannot have one process writing to it and another one reading from it at the same time. You need to open, read, deserialize, serialize, write, close.

That means that if you have an external process generating items to be blocked, every update needs to edit the file and then trigger the IPFS implementation to re-read it (possibly in full). Or the IPFS implementation needs to provide API to make the updates itself (and then write the file (again, possibly in full).

I have the feeling this format is not well suited to very large lists, or those updated frequently.

Reply:
Would defining a max list size as part of this spec help?
Together with `Link` field, it allows for composing big lists from smaller ones, and updates would only change part of a bigger DAG.

Reply:
That doesn't improve anything if instead of 1 item of 100GB we have 100 items of GB, somehow linked.

I also don't think the list format should include a Link field. Links to multiple lists etc should be a meta-format above that can be implemented later if we need some glue to handle multi-subscriptions (right now not needed).

I personally think lists should be line-based files, as small as possible, editable by hand for simple things like appending a CID. See lists used by UBlock etc.

Reply:
## Inefficiencies of CBOR lists vs prior art from adblockers

Ack. A compact, efficient, line-based syntax similar to [static filter lists used by adblockers](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax) ([example](https://easylist.to/easylist/easylist.txt)) will be more efficient and easier to consume than having CBOR.


Main problem: duplicated  "Type" "Content" and "Description" strings for every entry – havign these strings over and over again in CBOR map keys gets very wasteful very fast if the list is big.

## Manifest vs Compact list

So.. this PR feels like the meta format for aggregating lists you describe.  In [the other comment](https://github.com/ipfs/specs/pull/340#discussion_r1026509423) @hsanjuan  suggested adding a type for already existing double hashed format from https://badbits.dwebops.pub, we could do the same here and specify a separate type for lists in the "compact syntax". 

I think we should decouple and work on these things in parallel (meta-manifest and the compact list formats).
Will make discussion more productive, and allow us to ship things faster (e.g., in Kubo, we need compact format sooner than the meta-manifest one: users should be able to self-service, will take time).

## Next steps

@hsanjuan i'd like to explore your proposal – would you have time to propose a syntax for compact list? 
(gist / hackmd is fine – I can turn it into a separate IPIP  if you have limited bandwidth – this way we could flesh it out without blocking this one, and then we can decide which approach makes more sense)

(side-question) @thibmeu how strong is the CBOR  requirement?

Reply:
`Entries` structure could be transformed to a CSV. That's also the reason why multihash is serialised to a string via multibase, so it's easier to export.
```
type,content,description,categories
1,/path/to/content,questionable content,1001 1002 1005
...
```
I agree this would be easier to update. Metadata around it could be passed as a frontmatter possibly
```
action: 1
categories: 1000 1001
description: an example content list
---
type,content,description,categories
1,/path/to/content,questionable content,1001 1002 1005
```

The front matter still allows for compostability with CBOR, but makes the list self contained in a text format if need be.

I would not worry about the existing format for bad bits list. It's a single list, and I would avoid requiring the spec and implementation to be constrained by one historical list instance (similar to Cloudflare safemode in a way).

I have no need for CBOR specifically, it's just a format supported by IPFS and to which an IPLD schema can serialise to.

Reply:
> i'd like to explore your proposal – would you have time to propose a syntax for compact list?

yes, next week at some point? I have ideas but I also need to explain them (because they link to other aspects that are not related to the list itself), which takes more time. Format would be in the direction @thibmeu proposes.